### ISSUE TITLE
How to control lifetime of result object with boost::asio::use_future

### ISSUE THREAD

**Issue by neogenie** [yandex/ozo#283]
If we use simple callback we can create unique or shared pointer for extend the lifetime of db result object like this:

```
auto ptr = std::make_unique<ozo::rows_of<...>>();
auto& result = *ptr;
ozo::request(provider, query, ozo::into(result), 
        [result=std::move(ptr)] (ozo::error_code ec, auto conn) {
    if (ec) {
        log_error(ec);
        return;
    }
    process_result(*result);
}
```

Then if we use future to get result how to control the lifetime of result object ?

```
auto future = ozo::request(provider, query, result, boost::asio::use_future);
```

**Comment by thed636** [comment]
Hi Neo!

Well, there is no difference with using callbacks - the result object should have the life-time that exceeds the operation life-time which terminates with `future.get()` unblocking moment. E.g.:

```c++
auto result = std::make_unique<ozo::rows_of<...>>();
auto future = ozo::request(provider, query, ozo::into(*result), boost::asio::use_future);

//some additional code

auto conn = future.get();
do_something(*result);
result.reset();
```

Hope that helps.

**Comment by neogenie** [comment]
It's clear but it complicates the use of the library significantly. Using use_future imho assumes that we can return it outside so that the calling code when executing a request does not receive the result of its execution itself synchronously, but a future, as is done in most frameworks. In this case, it is not clear how to store the buffer and control its lifetime.

What is the general point in passing a buffer for the result from outside to the library, and not allocating it internally if necessary and returning it by value?

Workaround is return not a future, but a custom wrapper object containing the future and the actual pointer to the buffer (to extend its lifetime). But it's seems more complicated...

**Comment by thed636** [comment]
Yes, you right, that would be nice to have a result right from the future object, but here is the approach's rationale.

The first - to use the universal interface that `boost::asio` provides various types of continuations, including the `boost::asio::coroutine` that emulates a stackless coroutine with the `switch`-based automata. So, in this case, it is preferable to have (almost) the same signature for all the continuations. And here we have the [Handler](https://yandex.github.io/ozo/interface_handler.html) concept with a signature `void(error_code, Connection)`.

The second is not to impose on users how a result object is allocated. So the result may be allocated on the stack if no dynamic allocation is needed. This approach gives freedom to the user. It is similar to [boost::asio::async_read](https://www.boost.org/doc/libs/1_75_0/doc/html/boost_asio/reference/async_read.html) where the user provides the buffer. 

And the third is that a single PostgreSQL type may be mapped to several C++ types, so this interface allows the user to choose which C++ type should represent the type from a database. So this gives another choice for the user.

The library's main idea is not to impose user how but give the facilities to solve problems. It may consume additional time to involve but gives more freedom for a solution.

**Comment by neogenie** [comment]
@thed636, Thanks!
